API Go module

Joerger opened the pull request:
Changes:
- Initialize the `api` module and import it directly into the main `teleport` module.
- Update `make update-vendor` to delete `api` from vendor, and add a symlink in its place.

Joerger commented:
@russjones @awly I'm not sure exactly what our vendoring strategy is. How do normally update dependencies, and are the upgrades here an issue?

a-palchikov reviewed (commented):
Regarding upgrades - these are better done intentionally, e.g. when specific issues are targeted. But regular upgrades are also possible, I guess with the caveat that things need to continue working afterwards.

russjones commented on api/go.mod:
```diff
@@ -0,0 +1,19 @@
+	github.com/gogo/protobuf v1.3.2
+	github.com/golang/protobuf v1.4.3
+	github.com/gravitational/trace v1.1.14
+	github.com/jonboulle/clockwork v0.2.2 // indirect
+	github.com/sirupsen/logrus v1.7.0 // indirect
```
I know it's indirect, but do you know what's still using `clockwork` and `logrus`?

Joerger replied:
Yes, it's from `gravitational/trace`

russjones commented:
@Joerger I'm not sure if we should vendor `api` here. @awly what do you think?

awly reviewed (commented):
See https://github.com/golang/go/issues/33789 for the vendoring problem.
Can you try adding some post-processing steps in `make update-vendor` to remove the vendored copy of `api/`?

awly commented:
See how consul does this: https://github.com/hashicorp/consul/blob/master/GNUmakefile#L225-L228

Commit:
mfa: replace u2f-host with github.com/flynn/u2f (#5477)


This change removes the need for users to manually install u2f-host.
It also enables us to do U2F authentication with multiple devices.

Commit:
Fix windows cross-compilation with CGO (#5485)

Install `mingw-w64` cross-compiler toolchain in the buildbox and pass
magic flags to `go build` to use it.

Joerger committed:
Initialize teleport/api module, vendored for main teleport module.

Joerger commented on api/go.mod:
```diff
@@ -0,0 +1,19 @@
+module github.com/gravitational/teleport/api
```
I'm going to come back to versioning in a separate PR

Joerger commented:
> See [golang/go#33789](https://github.com/golang/go/issues/33789) for the vendoring problem.
> Can you try adding some post-processing steps in `make update-vendor` to remove the vendored copy of `api/`?

Post-processing does not seem to be possible in go 1.15, it leads to an "inconsistent vendoring error":

```bash
go: inconsistent vendoring in /home/bjoerger/gravitational/teleport:
	github.com/gravitational/teleport/api@v0.0.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/gravitational/teleport/api: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
``

> See how consul does this: https://github.com/hashicorp/consul/blob/master/GNUmakefile#L225-L228

If you upgrade Consol to 1.14+ you get the same issue:

```bash
go: inconsistent vendoring in /home/bjoerger/examples/consul:
	github.com/hashicorp/consul/api@v1.8.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/hashicorp/consul/sdk@v0.7.0: is explicitly required in go.mod, but not marked as explicit in vendor/modules.txt
	github.com/hashicorp/consul/api: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
	github.com/hashicorp/consul/sdk: is replaced in go.mod, but not marked as replaced in vendor/modules.txt
	github.com/gophercloud/gophercloud@v0.1.0: is marked as explicit in vendor/modules.txt, but not explicitly required in go.mod
```

I spent a while trying to find a way around this in 1.15, but it seems that partial vendoring of modules just isn't supported, and manual modification of `vendor` doesn't seem intended. I couldn't find any other way around vendoring `api`.

awly commented:
@Joerger I got this working locally with a symlink:
```
.PHONY: update-vendor
update-vendor:
	go mod tidy
	go mod vendor
	rm -r vendor/github.com/gravitational/teleport/api
	ln -s $(shell readlink -f api) vendor/github.com/gravitational/teleport
```

The problem with keeping `api` vendored, is that the compiler looks at the version in `vendor/`. So if you make a change in `api/` and forget to re-vendor, the change won't be picked up.

stevenGravy committed:
update enterprise quick start role example to 5.0 version (#5492)

* update role example to 5.0 version

benarent committed:
New Teleport Demo Video (#5494)

klizhentas committed:
Updated language and reordered things (#5504)

webvictim committed:
Add linting for Helm charts with example values (#5495)

benarent committed:
YAML error when coping and pasting (#5515)

stevenGravy committed:
Add kubernetes_service option to teleport helm chart (#5490)

* Increment to version 0.0.10

* Add kubernetes_service option config

Joerger committed:
Add symlink to vendoring of api.

Joerger commented:
> @Joerger I got this working locally with a symlink:
> 
> ```
> .PHONY: update-vendor
> update-vendor:
> 	go mod tidy
> 	go mod vendor
> 	rm -r vendor/github.com/gravitational/teleport/api
> 	ln -s $(shell readlink -f api) vendor/github.com/gravitational/teleport
> ```

@awly I'm not sure how to extend this solution to work non-locally, in docker/drone in particular. Is this a good final solution? I can't get anything other than a symlink to have the desired behavior. Another option is to ditch vendoring and depend solely on go mod since that alone should guarentee build consistency, but you may have a better idea as to why we need vendoring. Does teleport need to build without internet?

Joerger committed:
Remove deprecated marshalers (#5454)

Commit:
mfa: support multiple U2F keys on CLI login (#5484)

After adding several U2F tokens with `tsh mfa add`, you can now `tsh
login` using any of those tokens.

Two caveats:

1. The MFA method you get prompted for on login depends on the
`second_factor` config field on the auth server. There isn't yet an
option to require _either_ TOTP or U2F yet, even if you have both kinds
registered.

2. Web logins still need updating.

Also a few small unrelated changes:
- remove u2f-host binary presence check and docs
- hide `tsh mfa` commands until the feature is complete

Joerger committed:
Merge branch 'master' into joerger/api-client-go-mod

Joerger committed:
Edit vendor modules.txt to use local paths instead of vendoring api.

Joerger committed:
Change api vendor symlink to relative link for docker.

Joerger commented:
> @Joerger I got this working locally with a symlink

I changed this to a relative symlink to get it to work in docker without relinking.

Joerger committed:
Merge branch 'master' into joerger/api-client-go-mod

Joerger committed:
Refactor uuid usage to remove import.

Joerger commented on api/types/authentication.go:
```diff
@@ -308,13 +307,13 @@ func (u *U2F) Check() error {
 // NewMFADevice creates a new MFADevice with the given name. Caller must set
 // the Device field in the returned MFADevice.
-func NewMFADevice(name string, addedAt time.Time) *MFADevice {
+func NewMFADevice(name, id string, addedAt time.Time) *MFADevice {
```
@awly please take a quick look at this small refactor to remove the uuid dependency.

awly replied:
LGTM

awly commented on lib/services/authentication.go:
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/auth/u2f"
 	"github.com/gravitational/teleport/lib/utils"
+	"github.com/pborman/uuid"
 
 	"github.com/gravitational/trace"
 	"golang.org/x/crypto/bcrypt"
```
nit: group this import below with non-gravitational imports

Joerger committed:
Merge branch 'master' into joerger/api-client-go-mod

Joerger committed:
Regroup uuid import.

Joerger committed:
Add comment to update-vendor make target.

russjones committed:
Merge branch 'master' into joerger/api-client-go-mod

russjones reviewed (approved):
Bot.
